error_writer.go

```diff
@@ -46,74 +56,96 @@ func NewErrorWriter(opts ...HandlerOption) *ErrorWriter {
-		writer.allContentTypes[streaming] = struct{}{}
 	}
 	if config.HandleGRPC {
 		writer.grpcContentTypes[grpcContentTypeDefault] = struct{}{}
-		writer.allContentTypes[grpcContentTypeDefault] = struct{}{}
 		for name := range config.Codecs {
 			ct := grpcContentTypeFromCodecName(false /* web */, name)
 			writer.grpcContentTypes[ct] = struct{}{}
-			writer.allContentTypes[ct] = struct{}{}
 		}
 	}
 	if config.HandleGRPCWeb {
 		writer.grpcWebContentTypes[grpcWebContentTypeDefault] = struct{}{}
-		writer.allContentTypes[grpcWebContentTypeDefault] = struct{}{}
 		for name := range config.Codecs {
 			ct := grpcContentTypeFromCodecName(true /* web */, name)
 			writer.grpcWebContentTypes[ct] = struct{}{}
-			writer.allContentTypes[ct] = struct{}{}
 		}
 	}
 	return writer
 }
 
+func (w *ErrorWriter) classifyRequest(request *http.Request) protocolType {
+	ctype := canonicalizeContentType(getHeaderCanonical(request.Header, headerContentType))
+	if _, ok := w.unaryConnectContentTypes[ctype]; ok {
+		return connectUnaryProtocol
+	}
+	if _, ok := w.streamingConnectContentTypes[ctype]; ok {
+		return connectStreamProtocol
+	}
+	if _, ok := w.grpcContentTypes[ctype]; ok {
+		return grpcProtocol
+	}
+	if _, ok := w.grpcWebContentTypes[ctype]; ok {
+		return grpcWebProtocol
+	}
+	// Check for Connect-Protocol-Version header or connect protocol query
+	// parameter to support connect GET requests.
+	if request.Method == http.MethodGet {
+		connectVersion := getHeaderCanonical(request.Header, connectProtocolVersion)
+		if connectVersion == connectProtocolVersion {
+			return connectUnaryProtocol
+		}
+		connectVersion = request.URL.Query().Get(connectUnaryConnectQueryParameter)
+		if connectVersion == connectUnaryConnectQueryValue {
+			return connectUnaryProtocol
+		}
+	}
+	return unknownProtocol
+}
+
 // IsSupported checks whether a request is using one of the ErrorWriter's
 // supported RPC protocols.
 func (w *ErrorWriter) IsSupported(request *http.Request) bool {
-	ctype := canonicalizeContentType(getHeaderCanonical(request.Header, headerContentType))
-	_, ok := w.allContentTypes[ctype]
-	return ok
+	return w.classifyRequest(request) != unknownProtocol
 }
 
 // Write an error, using the format appropriate for the RPC protocol in use.
 // Callers should first use IsSupported to verify that the request is using one
-// of the ErrorWriter's supported RPC protocols.
+// of the ErrorWriter's supported RPC protocols. If the protocol is unknown,
+// Write will send the error in the Connect unary format.
 //
 // Write does not read or close the request body.
 func (w *ErrorWriter) Write(response http.ResponseWriter, request *http.Request, err error) error {
 	ctype := canonicalizeContentType(getHeaderCanonical(request.Header, headerContentType))
-	if _, ok := w.unaryConnectContentTypes[ctype]; ok {
-		// Unary errors are always JSON.
+	switch protocolType := w.classifyRequest(request); protocolType {
+	case connectUnaryProtocol, unknownProtocol:
```

Comment:
Can we make this the `default` branch below, to clarify that we're now never going to hit the unknown protocol case?